Make refined type printing more source compatible

jan-pieter opened the pull request:
Fixes #12939 
Supersedes #13698

SethTisue commented:
on the previous PR, @tgodzik wondered if the formatting could be improved. I think it's out of scope for this PR. Maybe Dale's recent pretty-printing PR (not merged yet) will improve the situation

SethTisue commented:
marking as draft for now since there are some test failures. probably just need to update some expected outputs to match the new format?

jan-pieter commented:
> probably just need to update some expected outputs to match the new format?
Most of the failures were indeed due to some unexpected output. I've updated the expected output for those.

The _catchall_ case was actually being hit by the tests for capture checking. I lack knowledge about the implementation details to explain that. For now I've updated this case to return an empty string which makes the tests pass. Some example output with the previous implementation:
```
-- [E007] Type Mismatch Error: tests/neg-custom-args/captures/i15772.scala:20:49 ---------------------------------------
20 |    val boxed1 : (({*} C) => Unit) -> Unit = box1(c)  // error
   |                                             ^^^^^^^
   |Found:    {c} 
   |  ({*} 
   |    ({c} 
   |      C{
   |        AnnotatedType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),class C),CaptureAnnotation(BiMapped3802(Var3741{TermRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),module class caps$)),val *)}, elems = {TermRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),module class caps$)),val *)}),false)) 
   |          arg
   |        : {*} C
   |      }
   |    )
   |   -> Unit)
   | -> Unit
   |Required: (({*} C) => Unit) -> Unit
   |
   | longer explanation available when compiling with `-explain`
```

bishabosha commented on compiler/src/dotty/tools/dotc/printing/PlainPrinter.scala:
```diff
@@ -112,7 +112,13 @@ class PlainPrinter(_ctx: Context) extends Printer {
-    (refinementNameString(rt) ~ toTextRHS(rt.refinedInfo)).close
+    val keyword = rt.refinedInfo match {
+      case _: TypeRef => "val "
+      case _: ExprType | _: MethodType => "def "
```
there could also be PolyType, e.g. `def bar: Foo { def foo[T <: Any](t: T): String } = ???`

bishabosha replied:
for this case, `Foo` must be a class that already defines a polymorphic method

jan-pieter commented:
Thanks for your comments and suggestions @bishabosha. I've updated the code & tests accordingly. This also solves the issue with `AnnotatedType`s in the capture checking tests.

jan-pieter commented:
@SethTisue Shall I rebase on the main branch, squash the commits and force push this branch?

dwijnand commented:
Is it just me that's bothered that `C1{val x: Int}` isn't spaced out as `C1 { val x: Int }`?

jan-pieter commented:
No, I prefer the latter one as well. However we previously said it to be out of scope for this PR: https://github.com/lampepfl/dotty/pull/16303#issuecomment-1307525828 . This PR only adds keywords. Perhaps https://github.com/lampepfl/dotty/pull/16277 did bring some improvements already? Otherwise I propose to improve that in a separate PR.

jan-pieter commented:
@SethTisue It took a few tries, but now it's ready to be merged I think :wink:

SethTisue commented:
Yay!
